pkg/cloudsql/adapter.go

```diff
@@ -0,0 +1,296 @@
+package cloudsql
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3/v2"
+	"github.com/dbtuneai/agent/pkg/agent"
+	"github.com/dbtuneai/agent/pkg/guardrails"
+	"github.com/dbtuneai/agent/pkg/internal/parameters"
+	"github.com/dbtuneai/agent/pkg/metrics"
+	"github.com/dbtuneai/agent/pkg/pg"
+	pgPool "github.com/jackc/pgx/v5/pgxpool"
+	"google.golang.org/api/sqladmin/v1"
+)
+
+type CloudSQLAdapter struct {
+	agent.CommonAgent
+	PGDriver              *pgPool.Pool
+	State                 *State
+	CloudSQLConfig        Config
+	CloudMonitoringClient *CloudMonitoringClient
+	CloudSQLAdminClient   *SqlAdminClient
+	GuardrailSettings     *guardrails.Config
+}
+
+func CreateCloudSQLAdapter() (*CloudSQLAdapter, error) {
+	ctx := context.Background()
+	commonAgent := agent.CreateCommonAgent()
+
+	pgConfig, err := pg.ConfigFromViper(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	pgPool, err := pgPool.New(ctx, pgConfig.ConnectionURL)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := ConfigFromViper()
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := monitoring.NewMetricClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	sqladminService, err := sqladmin.NewService(ctx)
+	if err != nil {
+		log.Fatalf("Failed to connect to SQL Admin service: %v", err)
+	}
+
+	guardrailSettings, err := guardrails.ConfigFromViper(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate settings for guardrails %w", err)
+	}
+
+	c := &CloudSQLAdapter{
+		CommonAgent:    *commonAgent,
+		State:          &State{LastGuardrailCheck: time.Now()},
+		PGDriver:       pgPool,
+		CloudSQLConfig: config,
+		CloudMonitoringClient: &CloudMonitoringClient{
+			client: client,
+			ctx:    ctx,
+		},
+		CloudSQLAdminClient: &SqlAdminClient{
+			client: sqladminService,
+		},
+		GuardrailSettings: &guardrailSettings,
+	}
+
+	c.InitCollectors(c.Collectors())
+
+	return c, nil
+}
+
+func (adapter *CloudSQLAdapter) ApplyConfig(proposedConfig *agent.ProposedConfigResponse) error {
```

Comment:
In this function also apply the the current configuration, i.e. do we lose user changes. Seems like we don't fetch them and then apply?

Reply:
Ahh I see this happens in `ApplyFlags`, never mind